File: neetcode/1. arrays/50.sortAnArray.py
# Given an array of integers nums, sort the array in ascending order and return it.

# You must solve the problem without using any built-in functions in O(nlog(n)) time complexity 
# and with the smallest space complexity possible

from typing import List

class Solution:
    def sortArray(self, nums: List[int]) -> List[int]:
        def merge(arr, L, M, R):
            left = arr[L:M+1]
            right = arr[M+1:R+1]
            i = L
            j = 0
            k = 0
            while j < len(left) and k < len(right):
                if left[j] <= right[k]:
                    arr[i] = left[j]
                    j += 1
                else:
                    arr[i] = right[k]
                    k += 1
            while j < len(left):
                nums[i] = right[k]
                k += 1
                i += 1

        def mergeSort(arr, l, r):
            if l == r:
                return arr
            m = (l + r) // 2
            mergeSort(arr, l, m)
            mergeSort(arr, m + 1, r)
            merge(arr, l, m, r)
            return arr
        
        return mergeSort(nums, 0, len(nums) - 1)

File: neetcode/1. arrays/50.sortAnArray.cpp
// Given an array of integers nums, sort the array in ascending order and return it.

// You must solve the problem without using any built-in functions in O(nlog(n)) time complexity 
// and with the smallest space complexity possible

#include <iostream>
#include <vector>
using namespace std;

class Solution {
public:
  vector<int> sortArray(vector<int>& nums) {
    mergeSort(nums, 0, nums.size() - 1);
    return nums;
  }
private:
  void merge(vector<int>& arr, int L, int M, int R) {
    vector<int> left(arr.begin() + L, arr.begin() + M + 1);
    vector<int> right(arr.begin() + M + 1, arr.begin() + R + 1);
    int i = L;
    int j = 0;
    int k = 0;
    while (j < left.size() && k < right.size()) {
      if (left[j] <= right[k]) {
        arr[i] = left[j];
        j++;
      } else {
        arr[i] = right[k];
        k++;
      }
      i++;
    }
    while (j < left.size()) {
      arr[i] = left[j];
      j++;
      i++;
    }
    while (k < right.size()) {
      arr[i] = right[k];
      k++;
      i++;
    }
  }

  void mergeSort(vector<int>& arr, int l, int r) {
    if (l < r) {
      int m = l + (r - l) / 2;
      mergeSort(arr, l, m);
      mergeSort(arr, m + 1, r);
      merge(arr, l, m, r);
    }
  }
};
